sockeye/train.py

```diff
@@ -767,35 +820,36 @@ def train(args: argparse.Namespace) -> training.TrainState:
             min_epochs=args.min_num_epochs,
             max_epochs=args.max_num_epochs,
             update_interval=args.update_interval,
-            stop_training_on_decoder_failure=args.stop_training_on_decoder_failure
+            stop_training_on_decoder_failure=args.stop_training_on_decoder_failure,
+            using_horovod=using_horovod
         )
         if trainer_config.min_epochs is not None and trainer_config.max_epochs is not None:
             check_condition(trainer_config.min_epochs <= trainer_config.max_epochs,
                             "Minimum number of epochs must be smaller than maximum number of epochs")
 
-        # Fixed training schedule always runs for a set number of updates
-        if args.learning_rate_schedule:
-            trainer_config.min_updates = None
-            trainer_config.max_updates = sum(num_updates for (_, num_updates) in args.learning_rate_schedule)
-            trainer_config.max_num_checkpoint_not_improved = -1
-            trainer_config.min_samples = None
-            trainer_config.max_samples = None
-            trainer_config.min_epochs = None
-            trainer_config.max_epochs = None
-
         optimizer_config = create_optimizer_config(args)
         training_model.initialize(optimizer_config.initializer, ctx=context)
         if args.params is not None:  # load existing parameters if present
             training_model.load_parameters(filename=args.params,
                                            ctx=context,
-                                           allow_missing=args.allow_missing_params or model_config.lhuc)
+                                           allow_missing=args.allow_missing_params or model_config.lhuc,
+                                           cast_dtype=True)
         params = training_model.collect_params()
         # set grad_req for fixed params
         params = set_grad_req_for_fixed_params(config=model_config,
                                                params=params,
                                                fixed_param_names=args.fixed_param_names,
                                                fixed_param_strategy=args.fixed_param_strategy)
 
+        # When using Horovod, synchronize the parameter initialization point
+        # across all workers by broadcasting worker 0's values.  This is not
+        # required when resuming training as synchronized training states
+        # already exist.
+        if using_horovod and not resume_training:
+            for ctx in context:
+                with mx.Context(ctx):
+                    horovod_mpi.hvd.broadcast_parameters(params, root_rank=0)
+
         if args.dtype == C.DTYPE_FP16:
             training_model.cast(C.DTYPE_FP16)
```

Comment:
I don't know much about AMP. Should one skip casting when using amp?

Reply:
The [sample AMP recipe](https://mxnet.incubator.apache.org/versions/master/tutorials/amp/amp_tutorial.html) doesn't involve any casting, so AMP should take care of data types automatically.  Sockeye would generally run with either `--amp` or `--dtype float16` but not both.

That said, we still aren't 100% sure AMP is working properly due to SoftmaxOutput and only include it for testing.  We could drop it from this update and keep it in the sockeye_2_training branch.

Reply:
Thanks for the explanation! Let's just add a TODO here that we might need to condition casting on args.dtype AND args.amp